visionmetrics/key_value_pair/key_value_pair_eval_base.py

```diff
@@ -0,0 +1,160 @@
+                "key_trace": ["brand_sentiment", "contoso_specific", "sentiment"]
+            },
+            "brand_sentiment_contoso_specific_logo_bounding_box": {
+                "metric_name": SupportedKeyWiseMetric.Detection_MeanAveragePrecision,
+                "metric_args": {"box_format": "xyxy", "coords": "absolute"},
+                "preprocessor": <reference to detection preprocessing function; see example implementation in key_value_pair_eval.py>,
+                "key_trace": ["brand_sentiment", "contoso_specific", "logo_bounding_box"]
+            }
+        }
+    }
+    """
+    def __init__(self, key_metric_map: dict):
+        super().__init__()
+        if not isinstance(key_metric_map, dict):
+            raise ValueError("key_metric_map must be a dictionary mapping keys to their corresponding evaluation metrics.")
+        self.key_metric_map = key_metric_map
+
+        self.key_evaluator_map = {}
+        for key in key_metric_map:
+            # Validate metric_name and metric_args
+            try:
+                metric_name = key_metric_map[key]["metric_name"]
+                metric_args = key_metric_map[key]["metric_args"]
+            except KeyError:
+                raise ValueError(f"Each value in key_metric_map must be a dictionary with a 'metric_name' key for a supported metric in visionmetrics of the form <task_name>.<metric_name>"
+                                 f" and a 'metric_args' key for a dictionary of arguments to pass to the initialization function of that metric. Key '{key}' does not satisfy this rule.")
+            if metric_name not in [m.value for m in SupportedKeyWiseMetric]:
+                raise ValueError(f"Metric '{key_metric_map[key]['metric_name']}' is not supported. "
+                                 f"Each key's metric must be among the supported metrics: {', '.join([m.value for m in SupportedKeyWiseMetric])}.")
+
+            # Instantiate metric
+            try:
+                self.key_evaluator_map[key] = eval(f"{metric_name}(**metric_args)")
+            except Exception as e:
+                raise ValueError(f"Encountered error '{repr(e)}' when instantiating metric '{metric_name}' for key '{key}' with arguments '{metric_args}'.")
+
+    def update(self, predictions, targets):
+        """
+        Updates metrics for each key using all samples in predictions and targets.
+        Both predictions and targets should be dictionaries of the form {'<key>': <value>}, where <value> is in the format expected for the respective metric for that key.
+        Each sample in predictions and targets must have the same keys (though they do not have to have all the keys in the dataset).
+        """
+        for key in self.key_evaluator_map:
+            metric = self.key_evaluator_map[key]
+            metric_name = self.key_metric_map[key]["metric_name"]
+            key_predictions = []
```

Comment:
`key_trace = self.key_metric_map[key]["key_trace"]`
Then reuse this variable in below for loop

Reply:
Good catch, thanks!